linkerd/stack/src/connect.rs

```diff
@@ -0,0 +1,124 @@
+use crate::{layer, Service};
+use futures::prelude::*;
+use linkerd_error::Error;
+use std::task::{Context, Poll};
+use tokio::io::{AsyncRead, AsyncWrite};
+
+/// A helper `Service` that drops metadata from a `MakeConnection`
+#[derive(Clone, Debug)]
+pub struct WithoutConnectionMetadata<S>(S);
+
+/// A helper that coerces a `MakeConnection` into a `Service`
+#[derive(Clone, Debug)]
+pub struct MakeConnectionService<S>(S);
+
+/// A helper trait that models a `Service` that creates client connections.
+///
+/// Implementers should implement `Service` and not `MakeConnection`. `MakeConnection` should only
+/// be used by consumers of these services.
+pub trait MakeConnection<T> {
+    /// An I/O type that represents a connection to the remote endpoint.
+    type Connection: AsyncRead + AsyncWrite;
+
+    /// Metadata associated with the established connection.
+    type Metadata;
+
+    type Error: Into<Error>;
+
+    type Future: std::future::Future<
+        Output = Result<(Self::Connection, Self::Metadata), Self::Error>,
+    >;
+
+    /// Determines whether the connector is ready to establish a connection.
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>>;
+
+    /// Establishes a connection.
+    fn make_connection(&mut self, t: T) -> Self::Future;
+
+    /// Returns a new `Service` that drops the connection metadata from returned values.
+    fn without_connection_metadata(self) -> WithoutConnectionMetadata<Self>
+    where
+        Self: Sized,
+    {
+        WithoutConnectionMetadata(self)
+    }
+
+    /// Coerces a `MakeConnection` into a `Service`.
+    fn into_service(self) -> MakeConnectionService<Self>
+    where
+        Self: Sized,
+    {
+        MakeConnectionService(self)
+    }
+}
+
+impl<T, S, I, M> MakeConnection<T> for S
+where
+    S: Service<T, Response = (I, M)>,
+    S::Error: Into<linkerd_error::Error>,
+    I: tokio::io::AsyncRead + tokio::io::AsyncWrite,
```

Comment:
nit: `AsyncRead` and `AsyncWrite` are imported in this file:
```suggestion
    I: AsyncRead + AsyncWrite,
```